logstash-core/src/main/java/org/logstash/execution/Filter.java

```diff
@@ -0,0 +1,132 @@
+package org.logstash.execution;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import org.logstash.Event;
+
+/**
+ * A Filter is simply a mapping of {@link QueueReader} to a new {@link QueueReader}.
+ */
+public interface Filter extends LsPlugin {
+
+    QueueReader filter(QueueReader reader);
+
+    @LogstashPlugin(name = "mutate")
+    final class Mutate implements Filter {
+
+        private static final PluginConfigSpec<String> FIELD_CONFIG =
+            LsConfiguration.requiredStringSetting("field");
+
+        private static final PluginConfigSpec<String> VALUE_CONFIG =
+            LsConfiguration.requiredStringSetting("value");
+
+        private final String field;
+
+        private final String value;
+
+        /**
+         * Required Constructor Signature only taking a {@link LsConfiguration}.
+         * @param configuration Logstash Configuration
+         * @param context Logstash Context
+         */
+        public Mutate(final LsConfiguration configuration, final LsContext context) {
+            this.field = configuration.get(FIELD_CONFIG);
+            this.value = configuration.get(VALUE_CONFIG);
+        }
+
+        @Override
+        public QueueReader filter(final QueueReader reader) {
+            return new QueueReader() {
+                @Override
+                public long poll(final Event event) {
+                    final long seq = reader.poll(event);
+                    if (seq > -1L) {
+                        event.setField(field, value);
+                    }
+                    return seq;
+                }
+
+                @Override
+                public long poll(final Event event, final long millis) {
```

Comment:
@andrewvc 

> Batching is necessary for things like the ES filter, etc

Yea maybe, but the efficient way of doing it is also highly plugin dependent. As of right now, I think it's fair to say that there are no plugins out there that actively make use of the batched interface in a quantifiable way. Otherwise, I wouldn't have found https://github.com/elastic/logstash/pull/8428 by mere chance :)

=> I'm not really comfortable with defining a batched API like that (yet). I'd much rather build a few plugins and see what works in the real world than guessing the right approach here now tbh.

Reply:
I don't see how it's plugin dependent? The only reason filters don't batch today is that it isn't possible because they can only get one event at a time. I'd rather not box us in.

Reply:
BTW, how does the efficiency of https://github.com/original-brownbear/logstash/pull/1/files#diff-eb733fe2c6eb811410b0e0a88226263d , it should be garbage free, aside from a single new Event[] call per batch, which is pretty cheap given a reasonable batch size.

Also, looking forward to your feedback on the `SimpleFilter` API.

I'd rather we only expose these APIs than the low level one here. The low level one is so implementation dependent it boxes us in. I'm also not sure who would actually want to use them directly, they're much more finnicky, and I'm not sure how much they buy a programmer in any situation I can think of.